### PR TITLE
Remove password fields in Admin/Users view. Add a send reset password…

### PR DESCRIPTION
**Pull request by acreilly**
… email button.

#438 

This PR needs to be merged with [solidus_auth_devise PR #97](https://github.com/solidusio/solidus_auth_devise/pull/97)

### CODE REVIEW

**Comment by gmacdougall** [comment]
I'm concerned about the tight coupling between this change and the required change in solidus_auth_devise. We like to maintain compatibility between extension versions.

Should we consider a strategy where we detect what solidus_auth_devise supports and then provide the correct functionality. Perhaps with a deprecation warning as well.

There are also failing tests which need to be addressed.

**Comment by acreilly** [comment]
I fixed the initial failing specs, but the specs will continue to fail due to the changes in solidus_auth_devise not being available on master.

I'm not exactly sure how to go about what you suggested. Since this change here would only be on master, it shouldn't affect other solidus versions. The change that was made on solidus_auth_devise won't break anything for older versions since that route is not used anywhere and is technically broken if anyone tried to use it.

**Comment by BenMorganIO** [comment]
Personally, I think there should be both. If we go for the password reset, an admin won't be able to manually change another users password.

Often, when onboarding, an admin will add an employee via the admin, give them a role, and then set their password to something and ask the new employee to update their password for security. If we do both, we allow an admin to be able to create a new user and set their password and then send a reset password email.

Both cases are beneficial and I see a use for both of them. If you're an admin or the head honcho, you have the OSS and you're running it with your own database. Nothing is actually stopping an admin from going into the console and setting another users password. The least that we could do is make it easier for them via having a web interface for it. The only security issue here is if they have a poor password or left their session open.

Therefore, we could just ask the current user to provider _their_ password as confirmation and also allow them to send out reset password emails. We should probably also wrap this in a `can?`.

File: backend/app/views/spree/admin/users/_form.html.erb [hunk 1]
```diff
@@ -53,21 +53,4 @@
...
-
-
-    <% if f.object.respond_to?(:password_confirmation) %>
-      <%= f.field_container :password do %>
-        <%= f.label :password_confirmation %>
-        <%= f.password_field :password_confirmation, class: 'fullwidth' %>
-        <%= f.error_message_on :password_confirmation %>
-      <% end %>
-    <% end %>
-  </div>
```
Comment (tvdeyen) [inline backend/app/views/spree/admin/users/_form.html.erb]: I think we should keep these fields and only display them

```rb
if @user == try_spree_current_user
```

also, we should add a "current password" field. Then we should have addressed all security concerns.

File: backend/app/views/spree/admin/users/edit.html.erb [hunk 2]
```diff
@@ -23,9 +23,22 @@
         </div>
       <% end %>
     <% end %>
+
   </div>
 </fieldset>
 
+<fieldset id='forgot-password'>
```
Comment (tvdeyen) [inline backend/app/views/spree/admin/users/edit.html.erb]: We should add a `no-border-bottom` class. There are already too much borders on this page, also I don't not see any need for adding a DOM ID.

File: backend/app/views/spree/admin/users/edit.html.erb [hunk 3]
```diff
@@ -23,9 +23,22 @@
...
+  <legend><%= Spree.t(:forgot_password) %></legend>
+
+  <%= form_for [@admin, Spree.user_class.new], as: :spree_user, url: spree.admin_reset_password_path do |f| %>
+      <%= f.hidden_field :email, value: @user.email %>
+
+    <div class="filter-actions actions">
```
Comment (tvdeyen) [inline backend/app/views/spree/admin/users/edit.html.erb]: We should just add a `align-center` class and should not make use of the `filter-actions actions` classes, as these will add even more borders to this already overwhelming page.

File: backend/app/views/spree/admin/users/edit.html.erb (hunk 2 above)
Comment (tvdeyen) [inline backend/app/views/spree/admin/users/edit.html.erb]: We should hide this conditionally 

```rb
if @user != try_spree_current_user
```

That this is only visible for admins that don't manage their own password.

File: core/config/locales/en.yml [hunk 4]
```diff
@@ -1731,6 +1731,7 @@ en:
     reports: Reports
     resend: Resend
     reset_password: Reset my password
+    reset_user_password: Reset user password
```
Comment (tvdeyen) [inline core/config/locales/en.yml]: By using the human attribute name, like suggested above, we can remove this.

**Comment by tvdeyen** [comment]
@acreilly there is a rebase necessary. Would you mind to?

**Comment by tvdeyen** [comment]
> Since this change here would only be on master, it shouldn't affect other solidus versions. The change that was made on solidus_auth_devise won't break anything for older versions since that route is not used anywhere and is technically broken if anyone tried to use it.

@gmacdougall @acreilly is right about this. Do you have anything particular in mind why this or the solidus auth PR should not be able to be merged?

**Comment by seand7565** [comment]
Hi all! This change looks great but it looks like it's been needing work for a while. This looks like a good fix to our current oldest issue, #438 - are there any blockers in making the above changes (other than the version changes that have happened since this PR?)

File: backend/app/views/spree/admin/users/edit.html.erb (hunk 2 above)
Comment (BenMorganIO) [reply backend/app/views/spree/admin/users/edit.html.erb]: Why? That breaks consistency with the UI.

**Comment by kennyadsl** [comment]
@acreilly hey there! I'm taking a look at this PR. There are some changes requested by @tvdeyen still to be done, especially the one about keeping fields if you are that user, since at the moment the UX generated with this PR is not ideal:

- as an Admin I need to change my password
- I visit my user page in admin
- I send the reset password link to myself
- I click on the link in the email
- I'm redirected to the homepage without any message and I can't change my password

This happens because to perform the reset it's required not to be authenticated:

```text
Redirected to http://localhost:3000/
Filter chain halted as :require_no_authentication rendered or redirected
Completed 302 Found in 6ms (ActiveRecord: 0.6ms)
```

Having the old password fields there for current users could be a solution for this problem.

Also, I think this needs a rebase. If you want or do not have time I can continue this PR with what's missing.

**Comment by zabrador** [comment]
@kennyadsl this seems pretty stale, and I believe it was superseded by https://github.com/solidusio/solidus_auth_devise/pull/146. Can this now be closed?

**Comment by kennyadsl** [comment]
Yes, you are right, thanks!

**Pull request closed without merging**